extensions/ql-vscode/src/codeql-task-provider.ts

```diff
@@ -0,0 +1,97 @@
+  private sharedState: string | undefined;
+
+  constructor(private workspaceRoot: string, private cliServer: CodeQLCliServer) { }
+
+  public async provideTasks(): Promise<vscode.Task[]> {
+    return this.getTasks();
+  }
+
+  public resolveTask(_task: vscode.Task): vscode.Task | undefined {
+    const command: string = _task.definition.command;
+    if (command) {
+      const definition: CodeQLTaskDefinition = <any>_task.definition;
+      return this.getTask(definition.command, definition.commandArgs ? definition.commandArgs : [], definition);
+    }
+    return undefined;
+  }
+
+  private getTasks(): vscode.Task[] {
+    if (this.tasks !== undefined) {
+      return this.tasks;
+    }
+    const commands: string[] = ['resolve languages', 'version'];
+    // Each command can have some commandArgs.
+    const commandArgs: string[][] = [[]];
+
+    this.tasks = [];
+    commands.forEach(command => {
+      commandArgs.forEach(commandArgsGroup => {
+        this.tasks!.push(this.getTask(command, commandArgsGroup));
+      });
+    });
+    return this.tasks;
+  }
+
+  private getTask(command: string, commandArgs: string[], definition?: CodeQLTaskDefinition): vscode.Task {
+    if (definition === undefined) {
+      definition = {
+        type: CodeQLTaskProvider.CodeQLType,
+        command,
+        commandArgs
+      };
+    }
+    return new vscode.Task(definition, vscode.TaskScope.Workspace, `${command} ${commandArgs.join(' ')}`,
+      CodeQLTaskProvider.CodeQLType, new vscode.CustomExecution(async (): Promise<vscode.Pseudoterminal> => {
+        // When the task is executed, this callback will run. Here, we setup for running the task.
+        return new CodeQLTaskTerminal(this.workspaceRoot, this.cliServer, command, commandArgs, () => this.sharedState, (state: string) => this.sharedState = state);
+      }));
+  }
+}
+
+class CodeQLTaskTerminal implements vscode.Pseudoterminal {
+  private writeEmitter = new vscode.EventEmitter<string>();
+  onDidWrite: vscode.Event<string> = this.writeEmitter.event;
+  private closeEmitter = new vscode.EventEmitter<number>();
+  onDidClose?: vscode.Event<number> = this.closeEmitter.event;
+
+  private fileWatcher: vscode.FileSystemWatcher | undefined;
+
+  constructor(private workspaceRoot: string, private cliServer: CodeQLCliServer, private command: string, private commandArgs: string[], private getSharedState: () => string | undefined, private setSharedState: (state: string) => void) {
+  }
+
+  open(): void {
+    void this.runCli([this.command], this.commandArgs);
+  }
+
+  close(): void {
+    // The terminal has been closed. Shutdown the build.
+    if (this.fileWatcher) {
+      this.fileWatcher.dispose();
+    }
+  }
+
+  private async runCli(command: string[], commandArgs: string[]): Promise<void> {
+    this.writeEmitter.fire('Running CodeQL task\r\n');
+    const output = JSON.parse(await this.cliServer.runCodeQlCliCommand(command, commandArgs, ''));
```

Comment:
So, this will use the existing cli server, or restart it if it's not currently running? I think this works.

Definitely need to handle malformed output in case the process abruptly ends partway through. Just add a try-catch with a friendly error message.